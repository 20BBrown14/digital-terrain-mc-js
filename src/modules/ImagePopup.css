.image-popup-modal {
  height: auto;
}

.image-popup-previous-button {
  float: left;
  top: 50%;
}

.image-popup-next-button {
  top: 50%;
  float: right;
}
