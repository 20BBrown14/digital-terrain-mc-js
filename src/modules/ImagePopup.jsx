import React from 'react';
import PropTypes from 'prop-types';
import {
  Modal,
  Button,
  Carousel,
  Row,
  Col,
} from 'antd';
import './ImagePopup.css';
import {
  RightCircleTwoTone,
  LeftCircleTwoTone,
} from '@ant-design/icons';

const propTypes = {
  /* Whether the modal is visible or not */
  isModalVisible: PropTypes.bool.isRequired,
  /* Function to close the modal */
  closeModal: PropTypes.func.isRequired,
  /* Images to show in modal */
  images: PropTypes.arrayOf(PropTypes.string),
  /* Index to start image carousel on */
  carouselIndexStart: PropTypes.number,
};

const defaultProps = {
  images: [],
  carouselIndexStart: 0,
};

class ImagePopup extends React.Component {
  constructor(props) {
    super(props);

    this.carousel = undefined;

    this.createCarouselRef = this.createCarouselRef.bind(this);
  }

  componentDidUpdate() {
    const { carouselIndexStart } = this.props;
    if (this.carousel) {
      this.carousel.goTo(carouselIndexStart);
    }
  }

  createCarouselRef(node) {
    this.carousel = node;
  }

  render() {
    const {
      isModalVisible,
      closeModal,
      images,
      carouselIndexStart,
    } = this.props;
    const buildImageComponents = () => (
      images.map((image, index) => (
        /* eslint-disable-next-line react/no-array-index-key */
        <img src={image} alt="" key={`image-${index}`} />
      ))
    );

    return (
      <Modal
        title="Image Popup"
        centered
        visible={isModalVisible && !!images}
        onCancel={closeModal}
        footer={null}
        className="image-popup-modal"
        width="80%"
      >
        <Row>
          <Col lg={1} md={2} sm={2} xs={2}>
            <Button
              className="image-popup-previous-button"
              shape="circle"
              icon={<LeftCircleTwoTone />}
              onClick={() => { this.carousel.prev(); }}
            />
          </Col>
          <Col lg={22} md={20} sm={20} xs={20}>
            <Carousel
              dots={false}
              autoplay={false}
              ref={this.createCarouselRef}
              initialSlide={carouselIndexStart}
            >

              {buildImageComponents()}
            </Carousel>
          </Col>
          <Col lg={1} md={2} sm={2} xs={2}>
            <Button
              className="image-popup-next-button"
              shape="circle"
              icon={<RightCircleTwoTone />}
              onClick={() => { this.carousel.next(); }}
            />
          </Col>
        </Row>
      </Modal>
    );
  }
}

ImagePopup.propTypes = propTypes;
ImagePopup.defaultProps = defaultProps;

export default ImagePopup;
