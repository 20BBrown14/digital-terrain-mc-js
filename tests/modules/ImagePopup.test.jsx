import React from 'react';
import ImagePopup from '../../src/modules/ImagePopup';

describe('ImagePopup', () => {
  it('renders a default view', () => {
    const wrapper = render(<ImagePopup isModalVisible closeModal={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('when there are images to show', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(
        <ImagePopup
          isModalVisible
          closeModal={() => {}}
          images={[{}, {}, {}, {}]}
        />,
      );
    });
    it('has a modal', () => {
      expect(wrapper.find('Modal')).toHaveLength(1);
    });

    it('has a grid', () => {
      expect(wrapper.find('Row')).toHaveLength(1);
      expect(wrapper.find('Col')).toHaveLength(3);
    });

    it('has two buttons', () => {
      expect(wrapper.find('Button')).toHaveLength(2);
    });

    it('has a carousel', () => {
      expect(wrapper.find('Carousel')).toHaveLength(1);
    });

    it('has images', () => {
      expect(wrapper.find('img')).toHaveLength(4);
    });
  });

  describe('componentDidUpdate', () => {
    let wrapper;
    let mockCarouselRef;
    beforeEach(() => {
      wrapper = shallow(
        <ImagePopup
          isModalVisible
          closeModal={() => {}}
          images={[{}, {}, {}, {}]}
          carouselIndexStart={3}
        />,
      );
      mockCarouselRef = { goTo: jest.fn() };
    });

    it('calls goto if carousel ref is set', () => {
      wrapper.instance().carousel = mockCarouselRef;
      wrapper.instance().componentDidUpdate();
      expect(mockCarouselRef.goTo).toHaveBeenCalledTimes(1);
      expect(mockCarouselRef.goTo).toHaveBeenCalledWith(3);
    });
  });

  describe('createCarouselRef', () => {
    it('sets carousel ref', () => {
      const wrapper = shallow(<ImagePopup isModalVisible closeModal={() => {}} images={[{}, {}, {}, {}]} />);
      wrapper.instance().createCarouselRef('someRef');
      expect(wrapper.instance().carousel).toEqual('someRef');
    });
  });

  describe('carousel nav buttons', () => {
    let wrapper;
    let mockCarouselRef;
    beforeEach(() => {
      wrapper = shallow(
        <ImagePopup
          isModalVisible
          closeModal={() => {}}
          images={[{}, {}, {}, {}]}
          carouselIndexStart={3}
        />,
      );
      mockCarouselRef = { next: jest.fn(), prev: jest.fn() };
      wrapper.instance().carousel = mockCarouselRef;
    });

    it('calls previous when prev button is clicked', () => {
      wrapper.find('Button').at(0).simulate('click');
      expect(mockCarouselRef.prev).toHaveBeenCalledTimes(1);
    });

    it('calls next when next button is clicked', () => {
      wrapper.find('Button').at(1).simulate('click');
      expect(mockCarouselRef.next).toHaveBeenCalledTimes(1);
    });
  });
});
