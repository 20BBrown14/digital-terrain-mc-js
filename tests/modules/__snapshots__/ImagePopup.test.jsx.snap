// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImagePopup renders a default view 1`] = `null`;
